### PR TITLE
Dev/xygu/20220124/theme fixes

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): unoplatform/private#239

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

Theme updates are not properly propagated to all resources

## What is the new behavior?

Fixed two areas where the resources are not receiving the theme updates.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information
closes unoplatform/uno#6735
closes unoplatform/uno#6761
closes unoplatform/uno#7118
closes unoplatform/uno#7151

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//#define DEBUG_SET_RESOURCE_SOURCE
+#define DEBUG_SET_RESOURCE_SOURCE
```
Comment (Xiaoy312) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:1]: fixme

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 2]
```diff
@@ -424,6 +425,11 @@ internal void CopyFrom(ResourceDictionary source)
 			_mergedDictionaries.Clear();
 			_themeDictionaries?.Clear();
 
+			// In a foreign library that uses merged-dict with 'Source=...', there can be multiple instances of res-dict from one single xaml file.
+			// And, the instance referenced in App.xaml may not be the same one used in merged-dict of library-a\textbox.xaml.
+			// In order to ensure the theme updates covers all of them, we need to keep track of this source instance for theme updates before it is lost.
+			_sourceDictionary = new WeakReference<ResourceDictionary>(source);
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:431]: Use `WeakReferencePool.Rent()` instead, it'll reuse the self-defined weak reference in `ResourceDictionary`

File: src/Uno.UI/UI/Xaml/ResourceDictionary.cs [hunk 3]
```diff
@@ -556,6 +562,11 @@ internal void UpdateThemeBindings(ResourceUpdateReason updateReason)
 			{
 				mergedDict.UpdateThemeBindings(updateReason);
 			}
+
+			if (_sourceDictionary?.TryGetTarget(out var target) == true)
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/ResourceDictionary.cs:566]: Typed WeakReference uses generics which is slower than using a cast (for now at least). When changing to ManagedWeakReference:

```suggestion
			if (_sourceDictionary.Target is ResourceDictionary target)
```

**Review by jeromelaban — COMMENTED** [review]
Are you able to write a unit tests for this one?

**Comment by Xiaoy312** [comment]
@jeromelaban Unfortunate, I was not able to replicate the same scenario that happened with Uno.Material library.

**Comment by jeromelaban** [comment]
You may be able to assert for that specific change's behavior, even if the original scenario is not reproducible.

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**